File: lib/src/main/java/io/github/kory33/guardedqueries/core/utils/extensions/SetExtensions.java
package io.github.kory33.guardedqueries.core.utils.extensions;

import com.google.common.collect.ImmutableSet;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class SetExtensions {
    private SetExtensions() {
    }

    /**
     * Union of elements from two collections.
     */
    public static <T> ImmutableSet<T> union(
            final Collection<? extends T> collection1,
            final Collection<? extends T> collection2
    ) {
        return ImmutableSet.<T>builder()
                .addAll(collection1)
                .addAll(collection2)
                .build();
    }

    /**
     * Intersection of elements from two collections.
     */
    public static <T> ImmutableSet<T> intersection(
            final Collection<? extends T> collection1,
            final Collection<? extends T> collection2
    ) {
        final var set2 = ImmutableSet.copyOf(collection2);
        return ImmutableSet.copyOf(collection1.stream().filter(set2::contains).iterator());
    }

    /**
     * Check if two collections have any common elements.
     */
    public static boolean nontriviallyIntersects(
            final Collection<?> collection1,
            final Collection<?> collection2
    ) {
        final var set2 = ImmutableSet.copyOf(collection2);
        return collection1.stream().anyMatch(set2::contains);
    }

    /**
     * Check if two collections have no common elements.
     */
    public static boolean disjoint(
            final Collection<?> collection1,
            final Collection<?> collection2
    ) {
        return !nontriviallyIntersects(collection1, collection2);
    }

    /**
     * Set difference of elements from two collections.
     */
    public static <T> ImmutableSet<T> difference(
            final Collection<? extends T> collection1,
            final Collection<? extends T> collection2
    ) {
        final var set2 = ImmutableSet.copyOf(collection2);
        return ImmutableSet.copyOf(collection1.stream().filter(e -> !set2.contains(e)).iterator());
    }

    /**
     * Powerset of a set of elements from the given collection, lazily streamed.
     */
    public static <T> Stream<ImmutableSet<T>> powerset(final Collection<? extends T> collection) {
        // deduplicated ArrayList of elements
        final var arrayList = new ArrayList<>(ImmutableSet.copyOf(collection));
        final var setSize = arrayList.size();

        // all non-negative BigInteger less than this value represents a unique subset of the given collection
        final var upperLimit = BigInteger.ONE.shiftLeft(setSize);

        final Iterator<ImmutableSet<T>> iterator = new Iterator<ImmutableSet<T>>() {
            BigInteger currentIndex = BigInteger.ZERO;

            @Override
            public boolean hasNext() {
                // currentIndex < upperLimit
                return currentIndex.compareTo(upperLimit) < 0;
            }

            @Override
            public ImmutableSet<T> next() {
                final var subset = ImmutableSet.<T>copyOf(
                        IntStream.range(0, setSize)
                                .filter(i -> currentIndex.testBit(i))
                                .mapToObj(arrayList::get)
                                .iterator()
                );
                currentIndex = currentIndex.add(BigInteger.ONE);
                return subset;
            }
        };

        return IteratorExtensions.stream(iterator);
    }
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/utils/extensions/SetExtensions.scala
package io.github.kory33.guardedqueries.core.utils.extensions

import scala.annotation.tailrec
import scala.collection.View
import IterableExtensions.given

object SetExtensions {
  given Extension: AnyRef with
    extension [T](set: Set[T])
      /**
       * Powerset of a set of elements from the given collection, lazily iterated and not
       * memoized.
       */
      def powerset: Iterable[Set[T]] = {
        val orderedSet = set.toList

        // every non-negative BigInteger less than this value represents a unique subset of the given collection
        val upperLimit = BigInt(1) << set.size

        // View.unfold is backed by a lazy iterator (UnfoldIterator) and not memoized
        View.unfold(BigInt(0))((currentIndex: BigInt) => {
          if (currentIndex < upperLimit) {
            val subset = (0 until set.size).filter(currentIndex.testBit).map(orderedSet(_))
            Some((subset.toSet, currentIndex + 1))
          } else {
            None
          }
        })
      }

      /**
       * Saturate this set by repeatedly applying a generator function `generator` that
       * generates a collection of elements of type `T` from values of type `T`.
       *
       * More precisely, the returned set is the smallest set `S` such that <ol> <li>`S`
       * contains all elements from the initial set</li> <li>for every element `t` of `S`,
       * `generator.apply(t)` is contained in `S`</li> </ol>
       */
      def generateFromElementsUntilFixpoint(generator: T => Set[T]): Set[T] = {
        @tailrec def recurse(
          elementsGeneratedSoFar: Set[T],
          elementsNewlyGeneratedInPreviousIteration: Set[T]
        ): Set[T] = {
          if (elementsNewlyGeneratedInPreviousIteration.isEmpty) {
            // we have reached the least fixpoint above the input set
            elementsGeneratedSoFar
          } else {
            val newlyGeneratedElements: Set[T] = for {
              // assuming that the generator function is pure,
              // it is only meaningful to generate new elements
              // from elements that have been newly added to the set
              // in the previous iteration
              newElementToConsider <- elementsNewlyGeneratedInPreviousIteration
              generatedElement <- generator.apply(newElementToConsider)
              if !elementsGeneratedSoFar.contains(generatedElement)
            } yield generatedElement

            recurse(
              elementsGeneratedSoFar ++ newlyGeneratedElements,
              newlyGeneratedElements
            )
          }
        }

        recurse(set, set)
      }

      /**
       * Saturate a collection of elements of type `T` by repeatedly applying a generator
       * function `generator` that generates a collection of elements of type `T` from a
       * collection of values of type `T`.
       *
       * More precisely, the returned set is the smallest set `S` such that <ol> <li>`S`
       * contains all elements from the initial collection</li> <li>`S union generator.apply(S)`
       * equals `S`</li> </ol>
       */
      def generateFromSetUntilFixpoint(generator: Set[T] => Set[T]): Set[T] = {
        @tailrec def recurse(elementsGeneratedSoFar: Set[T]): Set[T] = {
          val newlyGenerated = generator(elementsGeneratedSoFar)
          if (newlyGenerated.subsetOf(elementsGeneratedSoFar)) {
            // we have reached the least fixpoint above set
            elementsGeneratedSoFar
          } else {
            recurse(elementsGeneratedSoFar ++ newlyGenerated)
          }
        }

        recurse(set)
      }

      def widen[U >: T]: Set[U] = set.map(t => t)

      /**
       * Returns the [[Iterable]] that enumerates all sequences (with length `n`) of elements
       * from this set.
       */
      def naturalPowerTo(n: Int): Iterable[List[T]] = (0 until n).productAll(_ => set)

      /**
       * Union of all elements in this set, where each element is an iterable.
       */
      def unionAll[U](using ev: T <:< Iterable[U]): Set[U] = set.flatMap(ev(_))
}
